### ISSUE TITLE
add custom code to tp_traverse and tp_clear?

### ISSUE THREAD

**Issue by mdavidsaver** [cython/cython#2737]
I've been looking at Cython with an eye toward replacing some hand written
cpython extension code binding a C++ library.  This library is for a network
client and server, and includes a number of callbacks from C++ into python code.
It also makes heavy use of std::shared_ptr.

I find myself wanting the ability to add custom code to tp_traverse and tp_clear
to find additional python objects referenced indirectly through c++ objects.

I don't think this is presently possible?

A strawman example of the situation is as follows.
The reference cycles to be broken would include.

Holder.iface -> PyImpl.held -> Held -> ... -> Holder

```py
cdef extern from "foo.h":
    cdef cppclass Interface:
        ...
        void callback() # pure virtual
    cdef cppclass PyImpl(Interface):
        ...
        void callback() # calls held.callback()
        object held

    ctypedef Interface* pInterface

cdef class Held:
    def callback(self, arg):
        pass

cdef class Holder:
    cdef shared_ptr[Interface] iface

    # I would like to be able to do something like this:
    def tp_traverse(self, visit):
        cdef PyImpl* impl
        if self.iface:
            impl = dynamic_cast[pInterface](self.iface.get())
            if impl:
                visit(impl.held)
```

**Comment by scoder** [comment]
Could probably be added as two new special methods that get called by `tp_traverse` and `tp_clear`, so I'll leave this open as a feature request.

You can already do this by manually casting the type to a `PyTypeObject` and setting its slot methods. That way, you can store away the old slot value and insert your own one that calls the original one.

**Comment by mdavidsaver** [comment]
As advertised, it is possible to replace tp_traverse.  A working example: https://github.com/mdavidsaver/cython-c--demo

I found that the definition of `cpython.traverseproc` changed at some point to include "except -1".  In exploring ways to handle with this sort of API change I was surprised to find that eg. `IF CYTHON_HEX_VERSION>=0x001D01F0` isn't possible.

**Comment by scoder** [comment]
The example looks good.

> `IF CYTHON_HEX_VERSION>=0x001D01F0` isn't possible

Ah, yes, that makes sense. There aren't many real use cases for translation time environment tests, but this one is reasonable (although your specific use case is clearly better served by a cast). PR welcome that sets `CYTHON_HEX_VERSION` to what we now write into the C file.

**Comment by jdemeyer** [comment]
SageMath also does this, for example see
https://github.com/sagemath/sage/blob/develop/src/sage/structure/coerce_dict.pyx#L938

**Comment by navytux** [comment]
I also hit the need for this feature.

And I've made an observation that for `PyObject*` fields Cython also does not automatically generate `tp_traverse` and `tp_clear` methods:

---- 8< ---- `x.pyx`
```pyx
from cpython cimport PyObject

cdef class MyClass:
    cdef PyObject* linked_obj
```

---- 8< ---- `x.c`
```c
/* "x.pyx":3 
 * from cpython cimport PyObject
 * 
 * cdef class MyClass:             # <<<<<<<<<<<<<<
 *     cdef PyObject* linked_obj
 */
struct __pyx_obj_1x_MyClass {
  PyObject_HEAD
  PyObject *linked_obj;
};

...

static PyTypeObject __pyx_type_1x_MyClass = {
  PyVarObject_HEAD_INIT(0, 0)
  "x.""MyClass", /*tp_name*/
  sizeof(struct __pyx_obj_1x_MyClass), /*tp_basicsize*/
  ...
  0, /*tp_traverse*/
  0, /*tp_clear*/
  ...
};
```

Here `tp_traverse` and `tp_clear` are NULL and the user is expected to manually provide their implementations as suggested above.

But in my understanding there are a lot of code that embeds `PyObject*` - not  `object`, while there is only a few examples of custom tp_traverse and tp_clear overrides.

So all those codes currently leak memory if the linked object can lead to cycle.

**Comment by scoder** [comment]
> And I've made an observation that for `PyObject*` fields Cython also does not automatically generate `tp_traverse` and `tp_clear` methods

That is intended. `PyObject*` is not a reference, it's a pointer. 

> So all those codes currently leak memory if the linked object can lead to cycle.

If it's relevant for reference counting or garbage collection, then `PyObject*` is the wrong choice. The pointer type is intended to *exclude* Python object pointers from reference counting. If users don't want to exclude (or "hide" or "borrow") a pointer to a Python object from Python's memory management, then they shouldn't use `PyObject*` for it.

**Comment by navytux** [comment]
@scoder, thanks for feedback. I see your point. Yes, when `PyObject*` is used it is not a python reference, but more is a C pointer. Still sometimes `PyObject*` needs to be used - for example when an object to refer to might be NULL. Then the user need to care to manually Py_INCREF/Py_DECREF and Py_CLEAR. My point was that in that case there is currently no way built into Cython to support cyclic garbage collection at all.

**Comment by synodriver** [comment]
As long as you are implementing custom container types in cython, you will encounter this. I hope Cython can consider adding this feature as a special method, e.g. ```__traverse__``` and ```__clear__```, and no more manual cast.

**Comment by axeldavy** [comment]
+1 for the feature.

There are very significant speedups and advantages to using non-cpython structures when managing lists of items (list -> vector, dict -> map), in addition to not requiring the gil when accessing them.

Using a decorator for the PyObject* containers that we would want to include in tp_traverse/clear it not enough, as it would cause safety issues when tp_clear is run while the another thread uses the data without holding the gil. It needs to be a special method to properly handle user locking.

**Comment by da-woods** [comment]
I am broadly in favour of providing this because it is reasonable for classes to hold Python objects not directly through Cython (e.g. C++ containers). I think the main barrier is just whether this gets to the top of anyone's list of things to implement, rather than the the principle of the idea.

One question is whether it should override Cython's own default `tp_clear` and `tp_traverse` or add to them. I can see both variations being useful, and so it'd be good to have a nice was of doing this. Maybe just `__clear__` and `__clear_ex__` or something like that?

**Comment by synodriver** [comment]
I agree, two methods seems reasonable, and user is free to use one of them depend on whether they want to control python attributes.

**Comment by scoder** [comment]
For simplicity, let's add two new special methods, `__traverse__` and `__clear__`, which should behave like `__cinit__` in that Cython would call them from `tp_traverse()` and `tp_clear()` respectively.

I suggest calling them `__ctraverse__` and `__cclear__` to distinguish them from the bare protocols. Their intention is not to replace `tp_traverse` and `tp_clear` but to add to them in case Python object references are hidden from Cython by C pointers/structures/containers.

At which point in the `tp_*` methods the call happens is probably not very relevant, so I propose doing it *before* traversing the regular Python attributes. In the case of `tp_clear`, that means that the object is still (mostly) intact and thus more usable than afterwards. If users really, really need to clear something in a different order, they can manually clear object fields. Cython uses `Py_CLEAR()` for the fields, thus ignoring already cleared ones.

For an implementation, look at how `__cinit__` is implemented and do something similar.
